### PR TITLE
Migrate publish to gh actions

### PR DESCRIPTION
**Pull request by vzaidman**
2/2 commit converting Circle CI to Github actions.

Removed Circle CI to ensure we don't try to publish a version twice.

* Added the "deploy" step that only runs for tag pushes
* It waits for testing and validation, and fails if either of them fails
* This PR uses `--dry-run` to ensure no publish happens unexpectedly. This arg will be removed in a subsequent PR.

Successful CI with no deployment step when a PR is updated-
https://github.com/facebook/metro/actions/runs/10269830075
Screenshot 2024-08-06 at 16 54 20

Successful **DRY RUN** deploy from a hotfix branch (works pretty much the same on main)-
https://github.com/facebook/metro/actions/runs/10269576329/job/28415382215
Screenshot 2024-08-06 at 16 49 21

Deploy fails for a wrongly formatted tag (used tag "bad-tag-name") with a clear error-
https://github.com/facebook/metro/actions/runs/10269509288/job/28415186326
Screenshot 2024-08-06 at 16 29 19

Deploy fails for a tag pushed from a branch **not following** the Metro's release branch naming conventions
Screenshot 2024-08-06 at 17 01 28

### CODE REVIEW

File: .github/scripts/publish.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+echo "Trying to publish the package to npm for tag $RAW_TAG_NAME"
+
+# Validate tag's format follows conventions eg v00.01.02 or v0.90.2-alpha.5
```
Comment (robhogan) [inline .github/scripts/publish.sh:9]: Leading zeroes aren't allowed in the version part of a semver tag: https://semver.org/#spec-item-2

I don't think our validation needs to be perfect here (the regex could get over-complicated if we try to mirror semver) but the comment shouldn't have `v00.01.02` as an example of a conventional tag.

File: .github/workflows/build-test-and-deploy.yml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    tags:
+# The job is triggered for any tag push. Tag format validation will be done
+# as part of the deploy job for clearer error reporting on tag formatting
+      - '**'
+
+# head_ref is per PR, so this ensures that updating the latest PR commit
+# will cancel the previous run of the workflow and trigger a new one
+concurrency:
+  group: "build-test-and-deploy-${{ github.head_ref }}"
+  cancel-in-progress: true
```
[the remark this answers is not in the document]
Comment (vzaidman) [reply .github/workflows/build-test-and-deploy.yml:16]: Good point. Updated to `github` and updated the comment

File: .github/workflows/build-test-and-deploy.yml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  deploy:
+    # runs only on tag pushes
+    if: ${{ github.ref_type == 'tag' }}
+    runs-on: ubuntu-latest
+    name: "Deploy"
+    needs: [run-js-checks, test]
+    steps:
+      - uses: actions/checkout@v4
+      - uses: ./.github/actions/yarn-install
+      - run: echo "//registry.npmjs.org/:_authToken=${{ secrets.NPM_TOKEN }}" >> ~/.npmrc
```
[the remark this answers is not in the document]
Comment (vzaidman) [reply .github/workflows/build-test-and-deploy.yml:68]: As opposed to Circle CI the GH Actions default to be available to this top level yml only. Any secrets that are needed in actions (such as `actions/checkout@v4` or even `./.github/actions/yarn-install` above) has to be passed explicitly by [passing them as action input](https://docs.github.com/en/actions/security-for-github-actions/security-guides/using-secrets-in-github-actions#using-secrets-in-a-workflow) or by passing [`using secrets: inherit`](https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-inputs-and-secrets-in-a-reusable-workflow) to the job

**Review by robhogan — APPROVED** [review]

**Comment by robhogan** [comment]
LGTM, thanks!

**Pull request closed without merging**